### PR TITLE
introduction-to-processors

### LINKED ISSUE #919: An Introduction to Processors
Brief Summary:
In-depth knowledge on how a processor works is essential to engineering computers. While writing my approved topic on the x86 Processor, I found it necessary to include an initial article which introduces the reader to processors. I intend to provide a user-friendly guiding system on the structure of processors by introducing the reader to processors then using the architecture of the x86 Microprocessor to enhance their understanding.

Key Takeaways:
The objective of this article is to:
1. Increase understanding of how a CPU actually works.
2. Appreciate the use of the processor in real life applications.

References:
N/A

### PR DESCRIPTION
**Pull request by WangechiMurimi**
Fixes #919 
#### Preliminary Checks - Formatting and Structure 
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**
- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**
- [ ] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
- [x] In your PR - add a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.

**NOTE: (1st time contributors - submit ONLY one article to get familiar with the process and style-once published, you can then submit any number of articles.)**
      
#### Grammar Checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) -  [Grammarly](https://grammarly.com) 
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical Checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**

### Contribution Guidelines
For first-time contributors and for more details, **see [Contributing Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**

### CODE REVIEW

File: articles/introduction-to-processors/index.md [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+bb    01 00 00 00   #moving "stdout" number to ebx
+b8    04 00 00 00   #moving "print out" syscall number to eax
+cd    80            #calling the linux kernel to execute our print to stdout
+b8    01 00 00 00   #moving "sys_exit" call number to eax
+cd    80            #executing it via linux sys_call
+```
+
+
+
+To understand how code is turned from human-readable to machine-readable then executed, we shall use the __x86 Instruction Set Architecture__. This refers to a series of microprocessors developed by Intel based on the 8085 Processor. A __microprocessor__ is a programmable electronic chip that has computing and decision-making capabilities. These microprocessors(x86) share similar instruction set architectural models and are hence grouped. For example: 80186 & 80286 microprocessors.
```
Comment (lalith1403) [inline articles/introduction-to-processors/index.md:65]: Can you elaborate on the x86 instruction set?Additional information is required.

File: articles/introduction-to-processors/index.md [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+The 8086 is a 16-bit microprocessor that has 20 address lines and 16 data lines. An address bus is an internal channel that transfers the RAM location of the data being processed or the instruction being fetched. The number of lines (wires) in the address bus determines the maximum amount of RAM that the CPU can directly access as each line carries one bit of the address. Therefore, the x86 Processor has 2^(20) = 1,049,576 bytes = Approximately 1MB of RAM. 
+
+The 8086 microprocessor is divided into two functional parts: __Bus Interface Unit__(BIU) and the __Execution Unit__(EU). The Bus Interface Unit provides the interface of 8086 to external memory and I/O devices while the Execution Unit performs the operations and calculations as instructed by the computer program.
+
+
+The 8086 utilizes _pipelining_ to improve performance. This is a technique that balances the fetching and execution of an instruction, i.e., while an instruction is being executed, another is fetched. Pipelining increases the overall instruction throughput, thus speeds up the microprocessor.
```
Comment (lalith1403) [inline articles/introduction-to-processors/index.md:82]: Explain pipelining using a timing diagram.

File: articles/introduction-to-processors/index.md [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+The Execution Unit triggers the start of the Bus Interface Unit by sending instructions. After finding the location required, it is accessed, and data is retrieved using the data bus. The Execution Unit then removes instructions from the Instruction Stream Byte Queue for processing. The Bus Interface Unit simultaneously fetches more instructions from memory and puts them in the queue.
+
+In this article, we have outlined the basic functionality of a processor. In the next article, we will dive into the details using the x86 Processor’s Instruction Set Architecture.
```
Comment (lalith1403) [inline articles/introduction-to-processors/index.md:86]: Some more details in this article would add more value to the readers. Give an introduction here to the instruction set. In the next article, go in depth.

File: articles/introduction-to-processors/index.md [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+![alt text](x86TrainerKit.png)
+
+
+
+The figure above represents the 8086 trainer kit. A trainer kit is a device that functions as a simulator of the actual 8086 processor. It is used as a tool for learning, teaching, and development.
+
+The 8086 is a 16-bit microprocessor that has 20 address lines and 16 data lines. An address bus is an internal channel that transfers the RAM location of the data being processed or the instruction being fetched. The number of lines (wires) in the address bus determines the maximum amount of RAM that the CPU can directly access as each line carries one bit of the address. Therefore, the x86 Processor has 2^(20) = 1,049,576 bytes = Approximately 1MB of RAM. 
+
+The 8086 microprocessor is divided into two functional parts: __Bus Interface Unit__(BIU) and the __Execution Unit__(EU). The Bus Interface Unit provides the interface of 8086 to external memory and I/O devices while the Execution Unit performs the operations and calculations as instructed by the computer program.
```
Comment (lalith1403) [inline articles/introduction-to-processors/index.md:79]: Add an image here!

Comment (lalith1403) [reply articles/introduction-to-processors/index.md:79]: for both BIU and EU.

**Review by lalith1403 — CHANGES_REQUESTED** [review]
Amazing article. It is a good guide for beginners. Explain pipelining in detail, and focus on adding an introduction to the x86 instruction set.

**Comment by hectorkambow** [comment]
@WangechiMurimi  
👋 Good morning and thank you v much for submitting your 1st article to the EngEd Program.  

We're happy to have you join us.

Quick FYI - upon running your article through our 3rd party plagiarism checker it raised a few red flags and the % was higher than typically accept.

Please review the PDF and be sure to add sources and citations to the article as articles where needed.

Here is a link to help with any [confusion regarding plagiarism](https://www.scribbr.com/category/plagiarism/) - as we strive to create both informative and unique content for the community.

Note: The code should also be as original as possible - unless you are referencing specific frameworks or boilerplate needed to run the program (code snippet). 👍 

Below is the PDF.
[introduction-to-processors #1039.pdf](https://github.com/section-io/engineering-education/files/5748277/introduction-to-processors.1039.pdf)

Thank you 👍

**Comment by lalith1403** [comment]
@WangechiMurimi Do you plan to make changes to the article anytime soon? Let us know your timeline so that we can allocate resources accordingly.

**Comment by hectorkambow** [comment]
Just closing the article for now to clear up the queue @WangechiMurimi  
Can be re-opened at anytime to continue.

**Pull request closed without merging**